Add fixed size vectors

ahoarau opened the pull request:
This PR adds `eigen_vector2`,`eigen_vector3` and `eigen_vector4` so we can use them in scripting.
Implementation is a bit ugly, but I'd like also to do the same for matrix2d,3d,4d, the default in eigen.

To test it : 
```c 
import("eigen_typekit")

var eigen_vector2 a
var eigen_vector2 b = array(1.,2.)

var eigen_vector3 c
var eigen_vector3 d = array(1.,2.,3.)

var eigen_vector4 e
var eigen_vector4 f = array(1.,2.,3.,4.)
```
Gives : 

```
Deployer [S]> import("eigen_typekit")
 = true                

Deployer [S]> 
Deployer [S]> var eigen_vector2 a
 =                     0                   
0                   

Deployer [S]> var eigen_vector2 b = array(1.,2.)
0.905 [ Warning][Logger] Conversion from array to eigen_vector2
 =                     1                   
2                   

Deployer [S]> 
Deployer [S]> var eigen_vector3 c
 =                     0                   
0                   
3.84581e-315        

Deployer [S]> var eigen_vector3 d = array(1.,2.,3.)
0.910 [ Warning][Logger] Conversion from array to eigen_vector3
 =                     1                   
2                   
3                   

Deployer [S]> 
Deployer [S]> var eigen_vector4 e
 =                     9.75025e+199        
5.55604e+180        
1.27735e-152        
3.68066e+180        

Deployer [S]> var eigen_vector4 f = array(1.,2.,3.,4.)
2.325 [ Warning][Logger] Conversion from array to eigen_vector4
 =                     1                   
2                   
3                   
4     
```

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -72,46 +77,55 @@ namespace Eigen{
-        if (index >= (int) (v.size()) || index < 0)
-            return RTT::internal::NA<double>::na();
-        return v[index];
+#define DECLARE_VECTOR_GET_FUNCTIONS(VectorType) \
```
This section and the macro instantiations below can probably be replaced by template functions.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -157,17 +172,24 @@ namespace Eigen{
                 if ( id_name->get() == "size" || id_name->get() == "capacity") {
                     try {
-                        return RTT::internal::newFunctorDataSource(&get_size, RTT::internal::GenerateDataSource()(item.get()) );
+                        int (*get_size_pf)(const VectorType&) = get_size;
```
...which might also remove the need for these temporary function pointers.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -72,46 +77,55 @@ namespace Eigen{
         {
         };
         
+        const std::string type_name_; 
```
Please remove this member variable and use `getTypeName()` provided by the base class instead.

ahoarau commented:
Isn't there any way to "hide" the fixed-sized vectors into regular orocos eigen_vectors ?
Or maybe a simple way to define custom fixed sizes vectors/matrix in our own projects ?
(ex defining a twist would require a 6D vector..)

meyerj commented:
I would not know how to generate a typekit that supports a dynamic number of types or variable template arguments. The type system of RTT is based on the idea that each type can be used to look up a `TypeInfo` instance (if a typekit has been loaded) and the `TypeInfo` provides various type-specific methods that need to be compiled for each possible type or template instantiation.

What *is* possible is to expose the template declarations of `VectorTypeInfo<T>` and the other methods in a header file. This way you could register custom types in your own Orocos executable, another typekit plugin or in the constructor of a component that uses these types, for example:

```cpp
typedef Eigen::Vector<double, 6, 1> Vector6d;
using namespace RTT::types;
if (!Types()->getTypeInfo<Vector6d>()) {
  Types()->addType(new VectorTypeInfo<Vector6d>("eigen_vector6"));
}
```

The extern template declarations and explicit instantiations are not required and only an optimization on compile time and code size.

But the `VectorTypeInfo<T>` and related functions should not be exposed as part of the `Eigen` namespace and deserve their own namespace `eigen_typekit` or similar instead. There might be some difficulties related to [argument depending lookup](http://en.cppreference.com/w/cpp/language/adl). For example, the `istream` streaming operators should be defined in the `RTT::types` namespace so that they will be found by the implementation of [TypeStreamSelector<T>](https://github.com/orocos-toolchain/rtt/blob/master/rtt/types/TypeStreamSelector.hpp) without having to "pollute" the `Eigen` namespace. An alternative to define streaming operators is to specialize this template class for Eigen types and don't use the streaming operators at all.

That reminds me that I once created a patch for RTT based on Boost's [enable_if](http://www.boost.org/doc/libs/1_65_1/libs/core/doc/html/core/enable_if.html) and [operator type traits](http://www.boost.org/doc/libs/1_65_1/libs/type_traits/doc/html/boost_typetraits/category/value_traits/operators.html), that removed the need to specify explicitly whether streaming operators are defined for a certain type and also made the `istream` operator optional...

ahoarau commented:
Ok I took care of your comments, and also took the liberty of adding the Vector6d and Matrix6d to the set.
```ruby
import("eigen_typekit")

var eigen_vector2 a
var eigen_vector2 b = array(1.,2.)

var eigen_vector3 c
var eigen_vector3 d = array(1.,2.,3.)
c = array(5.,4.,3.,2.,1.,0.)


var eigen_vector4 e
var eigen_vector4 f = array(1.,2.,3.,4.)


var eigen_vector3 e1 = eigen_vector3(1,11)
var eigen_vector3 e2 = eigen_vector3(123,2)


var eigen_matrix m1 = eigen_matrix(10,11)
var eigen_matrix4 m2 = eigen_matrix4(123,2)
```

Don't you think it'd be better to initialize the vectors/matrices to zero when resizing ?

meyerj commented:
> Don't you think it'd be better to initialize the vectors/matrices to zero when resizing ?

Indeed, this might be helpful. Normally Eigen does not initialize elements of vectors and matrices for performance reasons. But in all cases where the constructors provided by Orocos typekits are invoked performance is probably not the primary concern.

Especially for variable-sized vectors and matrices it would be useful to initialize elements, or to provide alternative constructors that initialize all elements (like [vector_index_value_constructor<VectorType>](https://github.com/orocos/rtt_geometry/pull/26/files#diff-954d2e4bf9df14bbd83fd92cec92b55aR339) for Eigen vectors).

Without an operator to access matrix elements (see https://github.com/orocos/rtt_geometry/issues/23, or at least an unary index in the default storage order column-major) and without the implementation of at least the common operators for addition and multiplication, Eigen types are not very useful in Orocos scripting anyway.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -387,23 +415,51 @@ namespace Eigen{
+        RTT::types::Types()->type("eigen_vector2")->addConstructor(types::newConstructor(vector_index_constructor<Vector2d>()));
+        RTT::types::Types()->type("eigen_vector3")->addConstructor(types::newConstructor(vector_index_constructor<Vector3d>()));
+        RTT::types::Types()->type("eigen_vector4")->addConstructor(types::newConstructor(vector_index_constructor<Vector4d>()));
+        RTT::types::Types()->type("eigen_vector6")->addConstructor(types::newConstructor(vector_index_constructor<Vector6d>()));
```
It does not make sense to add `vector_index_constructor`s for fixed-size vectors. Actually the constructor functor should be called `vector_size_constructor`: it takes a single argument and returns a new vector of the given size.

Using those constructors with an argument different than the fixed size triggers a failed assertion:
```
Deployer [S]> eigen_vector2(2)
 =                     6.90771e-310        
1.54278e-316        

Deployer [S]> eigen_vector2(3)
deployer-gnulinux: /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:285: void Eigen::PlainObjectBase<Derived>::resize(Eigen::Index) [with Derived = Eigen::Matrix<double, 2, 1>; Eigen::Index = long int]: Assertion `((SizeAtCompileTime == Dynamic && (MaxSizeAtCompileTime==Dynamic || size<=MaxSizeAtCompileTime)) || SizeAtCompileTime == size) && size>=0' failed.
Aborted
```

ahoarau replied:
Ok I fixed it

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -292,87 +313,94 @@ namespace Eigen{
+        const VectorType& operator()(std::vector<double > values) const
         {
-            (*ptr)=VectorXd::Map(&values[0],values.size());
+            (*ptr)=VectorType::Map(&values[0],values.size());
```
Check size before: Construction a fixed-size vector from an array triggers either an assertion or even a segmentation fault if the sizes do not match:

```cpp
Deployer [S]> var eigen_vector2 x = array(1.0)
14.145 [ Warning][Logger] Conversion from array to eigen_vector2
deployer-gnulinux: /usr/include/eigen3/Eigen/src/Core/MapBase.h:144: Eigen::MapBase<Derived, 0>::MapBase(Eigen::MapBase<Derived, 0>::PointerType, Eigen::Index) [with Derived = Eigen::Map<Eigen::Matrix<double, 2, 1>, 0, Eigen::Stride<0, 0> >; Eigen::MapBase<Derived, 0>::PointerType = double*; Eigen::Index = long int]: Assertion `dataPtr == 0 || SizeAtCompileTime == Dynamic || SizeAtCompileTime == vecSize' failed.
Aborted

Deployer [S]> var eigen_vector2 v2 = array()
13.036 [ Warning][Logger] Conversion from array to eigen_vector2
Segmentation fault

```

ahoarau commented:
I addressed the issues, and added a fixed size vector value init : `eigen_vector2(123)` will setConstant to 123.

```typescript
import("eigen_typekit")

var eigen_vector2 a
var eigen_vector2 b = array(1.,2.)

var eigen_vector2 b2 = array(1.,2.,3.)

var eigen_vector3 c
var eigen_vector3 d = array(1.,2.,3.)

var eigen_vector4 e
var eigen_vector4 f = array(1.,2.,3.,4.)
```

What's bothering me is that `var eigen_vector2 b2 = array(1.,2.,3.)` will output : 
```
Deployer [S]> var eigen_vector2 b2 = array(1.,2.,3.)
3.809 [ ERROR  ][Logger] Cannot copy an std vector of size 3 into a fixed size vector of size 2
 =                     1                   
2                   
```
The array is copied anyway probably because it is trying to assing each value individualy with the "[]" operator (theory).

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -335,36 +335,79 @@ namespace Eigen{
         const VectorType& operator()(int size,double value ) const
         {
-            ptr->resize(size);
+            ptr->conservativeResizeLike(VectorType::Zero(size));
```
This line can be removed? The assignment below will implicitly resize the Eigen vector.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -335,36 +335,79 @@ namespace Eigen{
+            if(ptr->size() != values.size() && VectorType::RowsAtCompileTime == Eigen::Dynamic)
+            {
+                ptr->conservativeResize(values.size());
+            }
```
For the dynamic-sized vector there is no need to resize explicitly. The assignment below will resize the vector as needed and all values will be overwritten anyway.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -335,36 +335,79 @@ namespace Eigen{
+                log(Error) << "Cannot copy an std vector of size " << values.size()
+                           << " into a fixed size vector of size " << ptr->size()
+                           << endlog();
+                return *(ptr);
```
There is no way to let an Orocos operator fail. In any case it should return a well-defined value, e.g. `VectorType::Constant(RTT::internal::NA<double&>::na())`. At the moment the instance `ptr` owned by the constructor singleton is not overwritten, and therefore the result is either undefined or whatever fixed-size vector has been constructed before.

See also #34.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -138,7 +138,7 @@ namespace Eigen{
 	        typedef typename RTT::internal::AssignableDataSource<VectorType >::shared_ptr sh_ptr;
             sh_ptr asarg = RTT::internal::AssignableDataSource<VectorType >::narrow( arg.get() );
-                asarg->set().resize( size );
+                asarg->set().conservativeResizeLike(VectorType::Zero(size));
```
:+1:

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -375,7 +418,7 @@ namespace Eigen{
         const VectorType& operator()(int size ) const
         {
-            ptr->resize(size);
+            ptr->conservativeResizeLike(VectorType::Zero(size));
```
`*ptr = VectorType::Zero(size)` ?

No need to keep the previous values in `ptr`. Over its lifetime it will only be resized and copied, so all elements are already zero.

meyerj commented on eigen_typekit/eigen_typekit.cpp:
```diff
@@ -402,11 +445,11 @@ namespace Eigen{
         const MatrixType& operator()(int size1,int size2) const
         {
             ptr->resize(size1,size2);
+            ptr->setZero();
```
Simpler:
`*ptr = MatrixType::Zero(size1,size2);`

ahoarau commented:
I took care of the comments. Now I'm wondering how to set to zero all the elements of the fixed-size vectors.
Could'nt manage to pass a function to `RTT::types::Types()->type("eigen_vector2")->addConstructor(types::newConstructor(vector_fixed_value_constructor<Vector2d>()));`.
Or create an unary function with a void argument.
